signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+            try:
+                with self._suspend_sync():
+                    for key in data:
+                        if key in self._data:
+                            if data[key] == self._data[key]:
+                                continue
+                            try:
+                                self._data[key].reset(key)
+                                continue
+                            except (ValueError, AttributeError):
+                                pass
+                        self._data[key] = self.from_base(data[key])
+                    remove = set()
+                    for key in self._data:
+                        if key not in data:
+                            remove.add(key)
+                    for key in remove:
+                        del self._data[key]
+                self.sync()
+            except BaseException:  # rollback
+                self._data = backup
+                raise
+        else:
+            raise ValueError("The data must be a mapping or None.")
+
+    @classmethod
+    def _validate_key(cls, key):
+        "Emit a warning or raise an exception if key is invalid. Returns key."
+        if isinstance(key, str):
+            if '.' in key:
+                from ..errors import InvalidKeyError
+                raise InvalidKeyError(
+                    "keys may not contain dots ('.'): {}".format(key))
+            else:
+                return key
+        elif isinstance(key, cls.VALID_KEY_TYPES):
+            return cls._validate_key(str(key))
+        else:
+            from ..errors import KeyTypeError
+            raise KeyTypeError(
+                "keys must be str, int, bool or None, not {}".format(type(key).__name__))
+
+    def __delitem__(self, item):
+        self.load()
+        del self._data[item]
+        self.sync()
+
+    def __setitem__(self, key, value):
+        self.load()
+        with self._suspend_sync():
+            self._data[self._validate_key(key)] = self.from_base(value, parent=self)
+        self.sync()
+
+    def __getitem__(self, key):
+        self.load()
+        return self._data[key]
+
+    def __iter__(self):
+        self.load()
+        return iter(self._data)
+
+    def __call__(self):
+        self.load()
+        return self.to_base()
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def keys(self):
+        self.load()
+        return self._data.keys()
+
+    def values(self):
+        self.load()
+        return self.to_base().values()
+
+    def items(self):
+        self.load()
+        return self.to_base().items()
+
+    def get(self, key, default=None):
+        self.load()
+        return self._data.get(key, default)
+
+    def pop(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
```

Comment:
Do we care about specifying the behavior if an error occurs during synchronization? Should it be able to roll back any changes to `_data`? (This may be less important than performance.)

Reply:
It fails while creating an instance of a child of `SyncedDict`.I don't need to add the data members of the child in `_Protected_key` as long as they are assigned before the call of `super().__init__()`.

Reply:
I think it's useful, or at least a feature we could consider including that could be turned on or off with a flag. Try/Except blocks are cheap as long as they don't trigger exceptions frequently.

Reply:
I see that `reset` has rollback that catches `BaseException`, so that's a pretty broad safety measure that seems like we'd want to mirror safety here.

Reply:
Okay, will add that.